File: SPV.CPP/SDK/Plugin/Transaction/Payload/WithdrawFromSideChain.cpp
// Copyright (c) 2012-2018 The Elastos Open Source Project
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "WithdrawFromSideChain.h"

#include <Common/Log.h>
#include <Common/Utils.h>

namespace Elastos {
	namespace ElaWallet {

		WithdrawFromSideChain::WithdrawFromSideChain() :
				_blockHeight(0),
				_genesisBlockAddress("") {

		}

		WithdrawFromSideChain::WithdrawFromSideChain(const WithdrawFromSideChain &payload) {
			operator=(payload);
		}

		WithdrawFromSideChain::WithdrawFromSideChain(uint32_t blockHeight, const std::string &genesisBlockAddress,
		                                           const std::vector<uint256> &sideChainTransactionHash) {
			_blockHeight = blockHeight;
			_genesisBlockAddress = genesisBlockAddress;
			_sideChainTransactionHash = sideChainTransactionHash;
		}

		WithdrawFromSideChain::~WithdrawFromSideChain() {
		}

		void WithdrawFromSideChain::SetBlockHeight(uint32_t blockHeight) {
			_blockHeight = blockHeight;
		}

		uint32_t WithdrawFromSideChain::GetBlockHeight() const {
			return _blockHeight;
		}

		void WithdrawFromSideChain::SetGenesisBlockAddress(const std::string &genesisBlockAddress) {
			_genesisBlockAddress = genesisBlockAddress;
		}

		const std::string &WithdrawFromSideChain::GetGenesisBlockAddress() const {
			return _genesisBlockAddress;
		}

		void WithdrawFromSideChain::SetSideChainTransacitonHash(const std::vector<uint256> &sideChainTransactionHash) {
			_sideChainTransactionHash = sideChainTransactionHash;
		}

		const std::vector<uint256> &WithdrawFromSideChain::GetSideChainTransacitonHash() const {
			return _sideChainTransactionHash;
		}

		size_t WithdrawFromSideChain::EstimateSize(uint8_t version) const {
			size_t size = 0;
			ByteStream stream;

			size += sizeof(_blockHeight);
			size += stream.WriteVarUint(_genesisBlockAddress.size());
			size += _genesisBlockAddress.size();
			size += stream.WriteVarUint(_sideChainTransactionHash.size());

			for (size_t i = 0; i < _sideChainTransactionHash.size(); ++i)
				size += _sideChainTransactionHash[i].size();

			return size;
		}

		void WithdrawFromSideChain::Serialize(ByteStream &ostream, uint8_t version) const {
			ostream.WriteUint32(_blockHeight);
			ostream.WriteVarString(_genesisBlockAddress);
			ostream.WriteVarUint((uint64_t)_sideChainTransactionHash.size());

			for (size_t i = 0; i < _sideChainTransactionHash.size(); ++i) {
				ostream.WriteBytes(_sideChainTransactionHash[i]);
			}
		}

		bool WithdrawFromSideChain::Deserialize(const ByteStream &istream, uint8_t version) {
			if (!istream.ReadUint32(_blockHeight)) {
				Log::error("Payload with draw asset deserialize block height fail");
				return false;
			}

			if (!istream.ReadVarString(_genesisBlockAddress)) {
				Log::error("Payload with draw asset deserialize genesis block address fail");
				return false;
			}

			uint64_t len = 0;
			if (!istream.ReadVarUint(len)) {
				Log::error("Payload with draw asset deserialize side chain tx hash len fail");
				return false;
			}

			_sideChainTransactionHash.resize(len);
			for (uint64_t i = 0; i < len; ++i) {
				if (!istream.ReadBytes(_sideChainTransactionHash[i])) {
					Log::error("Payload with draw asset deserialize side chain tx hash[{}] fail", i);
					return false;
				}
			}

			return true;
		}

		nlohmann::json WithdrawFromSideChain::ToJson(uint8_t version) const {
			nlohmann::json j;

			j["BlockHeight"] = _blockHeight;
			j["GenesisBlockAddress"] = _genesisBlockAddress;
			nlohmann::json hashes;
			for (size_t i = 0; i < _sideChainTransactionHash.size(); ++i)
				hashes.push_back(_sideChainTransactionHash[i].GetHex());
			j["SideChainTransactionHash"] = hashes;

			return j;
		}

		void WithdrawFromSideChain::FromJson(const nlohmann::json &j, uint8_t version) {
			_blockHeight = j["BlockHeight"].get<uint32_t>();
			_genesisBlockAddress = j["GenesisBlockAddress"].get<std::string>();

			nlohmann::json hashes = j["SideChainTransactionHash"];
			for (nlohmann::json::iterator it = hashes.begin(); it != hashes.end(); ++it)
				_sideChainTransactionHash.emplace_back((*it).get<std::string>());
		}

		IPayload &WithdrawFromSideChain::operator=(const IPayload &payload) {
			try {
				const WithdrawFromSideChain &payloadWithDrawAsset = dynamic_cast<const WithdrawFromSideChain &>(payload);
				operator=(payloadWithDrawAsset);
			} catch (const std::bad_cast &e) {
				Log::error("payload is not instance of WithdrawFromSideChain");
			}

			return *this;
		}

		WithdrawFromSideChain &WithdrawFromSideChain::operator=(const WithdrawFromSideChain &payload) {
			_blockHeight = payload._blockHeight;
			_genesisBlockAddress = payload._genesisBlockAddress;
			_sideChainTransactionHash = payload._sideChainTransactionHash;

			return *this;
		}

	}
}

File: ELA.CR.ClaimDPosNode/core/types/payload/withdrawfromsidechain.go
// Copyright (c) 2017-2020 The Elastos Foundation
// Use of this source code is governed by an MIT
// license that can be found in the LICENSE file.
// 

package payload

import (
	"bytes"
	"errors"
	"io"

	"github.com/elastos/Elastos.ELA/common"
)

const WithdrawFromSideChainVersion byte = 0x00

type WithdrawFromSideChain struct {
	BlockHeight                uint32
	GenesisBlockAddress        string
	SideChainTransactionHashes []common.Uint256
}

func (t *WithdrawFromSideChain) Data(version byte) []byte {
	buf := new(bytes.Buffer)
	if err := t.Serialize(buf, version); err != nil {
		return []byte{0}
	}

	return buf.Bytes()
}

func (t *WithdrawFromSideChain) Serialize(w io.Writer, version byte) error {
	if err := common.WriteUint32(w, t.BlockHeight); err != nil {
		return errors.New("[WithdrawFromSideChain], BlockHeight serialize failed.")
	}
	if err := common.WriteVarString(w, t.GenesisBlockAddress); err != nil {
		return errors.New("[WithdrawFromSideChain], GenesisBlockAddress serialize failed.")
	}

	if err := common.WriteVarUint(w, uint64(len(t.SideChainTransactionHashes))); err != nil {
		return errors.New("[WithdrawFromSideChain], SideChainTransactionHashes count serialize failed")
	}

	for _, hash := range t.SideChainTransactionHashes {
		err := hash.Serialize(w)
		if err != nil {
			return errors.New("[WithdrawFromSideChain], SideChainTransactionHashes serialize failed")
		}
	}
	return nil
}

func (t *WithdrawFromSideChain) Deserialize(r io.Reader, version byte) error {
	height, err := common.ReadUint32(r)
	if err != nil {
		return errors.New("[WithdrawFromSideChain], BlockHeight deserialize failed.")
	}
	address, err := common.ReadVarString(r)
	if err != nil {
		return errors.New("[WithdrawFromSideChain], GenesisBlockAddress deserialize failed.")
	}

	count, err := common.ReadVarUint(r, 0)
	if err != nil {
		return errors.New("[WithdrawFromSideChain], SideChainTransactionHashes count deserialize failed")
	}

	t.SideChainTransactionHashes = make([]common.Uint256, 0)
	for i := uint64(0); i < count; i++ {
		var hash common.Uint256
		err := hash.Deserialize(r)
		if err != nil {
			return errors.New("[WithdrawFromSideChain], SideChainTransactionHash deserialize failed.")
		}
		t.SideChainTransactionHashes = append(t.SideChainTransactionHashes, hash)
	}

	t.BlockHeight = height
	t.GenesisBlockAddress = address

	return nil
}
